job-server/src/spark.jobserver/JobServerNamedObjects.scala

```diff
@@ -0,0 +1,121 @@
+package spark.jobserver
+
+import akka.actor.{ ActorRef, ActorSystem }
+import akka.util.Timeout
+import scala.concurrent.Await
+import scala.concurrent.Future
+import scala.concurrent.ExecutionContext
+import scala.concurrent.duration._
+import org.apache.spark.SparkContext
+import org.slf4j.LoggerFactory
+import spray.caching.{ LruCache, Cache }
+import spray.util._
+
+/**
+ * An implementation of [[NamedObjects]] API for the Job Server.
+ * Note that this contains code that executes on the same thread as the job.
+ * Uses spray caching for cache references to named objects and to
+ * avoid that the same object is created multiple times
+ */
+class JobServerNamedObjects(system: ActorSystem) extends NamedObjects {
+  val logger = LoggerFactory.getLogger(getClass)
+
+  implicit val ec: ExecutionContext = system.dispatcher
+
+  // we must store a reference to each NamedObject even though only its ID is used here
+  // this reference prevents the object from being GCed and cleaned by sparks ContextCleaner
+  // or some other GC for other types of objects
+  private val namesToObjects: Cache[NamedObject] = LruCache()
+
+  override def getOrElseCreate[O <: NamedObject](name: String, objGen: => O)
+                                (implicit timeout: Timeout = defaultTimeout,
+                                    persister: NamedObjectPersister[O]): O = {
+
+    val res : Option[O] = getExistingObject(name)
```

Comment:
I simplified it to:

    cachedOp(name, createObject(objGen, name)).await(timeout) match {
      case obj: O @unchecked => obj
      case NamedRDD(_, _)    => throw new IllegalArgumentException("Incorrect type for named object")
    }